client/state/should-dehydrate-query.ts

```diff
@@ -0,0 +1,31 @@
+import { Query } from 'react-query';
+
+type PersistencePredicate< T > = ( data: T ) => boolean;
+
+const hasPersistenceSetting = ( value: unknown ): value is boolean => {
+	return typeof value === 'boolean';
+};
+
+const hasPersistenceCustomSetting = (
```

Comment:
Naming nit: it's not immediately clear what's the difference between `hasPersistenceSetting` and `hasPersistenceCustomSetting`. Should we make it more clear that we're taking about a boolean value vs a custom function?

Reply:
I thought about

`hasSimplePersistence`

and

`hasPersistencePredicate`

WDYT?

It might not be a boolean in the future, who knows?

Reply:
Sounds good to me 👍 . Naming is a very minor problem in this case, since we don't expose it outside the module anyway. This also means that if we decide to rename it later, it won't have any consequences related to backwards compatibility. Feel free to proceed either way 👍

Reply:
I was able to remove these types from the function body completely and TypeScript is happy about that, able to infer everything on its own:
```ts
export const shouldDehydrateQuery = ( query: Query ): boolean => {
  if ( query.state.status !== 'success' ) {
    return false;
  }

  const shouldPersist = query.meta?.persist;

  if ( typeof shouldPersist === 'boolean' ) {
    return shouldPersist;
  }

  if ( typeof shouldPersist === 'function' ) {
    return shouldPersist( query.state.data );
  }

  return true;
};
```

Reply:
The idea with the type guards was to explicitly define what we expect. `typeof shouldPersist === 'function'` makes it blurry, meaning I cannot understand from a single look why is that a function.

Having `hasPersistencePredicate`, instead, allows a more deep understanding of what's happening. The same rationale as having a variable instead of a magic number applies.